# Practice Session 04: Basket analysis

Author: <font color="blue">Nil Tomàs Plans</font>

E-mail: <font color="blue">nil.tomas01@estudiant.upf.edu</font>

Date: <font color="blue">15/10/2024</font>

In [1]:

import numpy as np  
import matplotlib.pyplot as plt  
import pandas as pd  
import csv
import gzip

from apyori import apriori

# 1. Playing with apyori

In [2]:
# LEAVE AS-IS

def print_apyori_output (association_results, info=False, info_key=False):
    for relation_record in association_results:
        itemset = list(relation_record.items)
        
        # Consider only itemsets of two elements
        if len(itemset) > 1: 
        
            print("Rules involving itemset %s" % itemset)
            support = relation_record.support

            for rules in relation_record.ordered_statistics:
                antecedent = list(rules.items_base)
                consequent = list(rules.items_add)
                
                if info_key:
                    antecedent = [info.loc[x][info_key] for x in antecedent]
                    consequent = [info.loc[x][info_key] for x in consequent]
                
                confidence = rules.confidence
                lift = rules.lift

                print("%s => %s (support=%.4f, confidence=%.2f, lift=%.2f)" %
                      (antecedent, consequent, support, confidence, lift))
            print()

<font size="+1" color="red">Replace this cell with your own example of transactions (at least 20 transactions). Execute the apriori algorithm, in which you should obtain at least <strong>two</strong> rules of the form ['A', 'B'] => ['C'], i.e., at least two rules having a 2-itemset in the antecedent and a 1-itemset in the consequent. Modify the transactions until you obtain such rules.</font>

In [3]:


transactions_bike_shop = [
    ["road_bike", "sun_glasses", "culotte", "helmet", "maillot"],
    ["road_bike", "sun_glasses", "energy_bars", "helmet", "shoes"],
    ["road_bike", "sun_glasses", "maillot", "shoes"],
    ["road_bike", "sun_glasses", "energy_bars", "helmet", "water_bottle"],#
    ["road_bike", "potentiometer", "rain_jacket", "socks", "sweater"],
    ["road_bike", "GPS", "socks", "sweater"],
    
    ["btt_bike", "gloves", "helmet"],
    ["btt_bike", "energy_drink", "gloves", "water_bottle"],
    ["btt_bike", "sun_glasses", "gloves", "helmet", "water_bottle"],#
    ["btt_bike", "culotte", "energy_bars", "GPS", "maillot"],
    ["btt_bike", "energy_drink", "helmet", "socks"],
    ["btt_bike", "sun_glasses", "GPS", "helmet", "rain_jacket", "water_bottle"],#
    ["btt_bike", "gloves", "GPS", "helmet", "potentiometer"],
    ["btt_bike",  "energy_drink", "helmet", "rain_jacket", "socks"],
    ["btt_bike", "rain_jacket", "socks"],    
    
    [ "GPS", "helmet", "potentiometer"],
    [ "sun_glasses", "helmet", "water_bottle"],#
    [ "gloves", "maillot", "shoes", "socks"],
    [ "energy_drink", "maillot", "socks"],
    [ "energy_drink", "gloves", "helmet", "potentiometer"],
    
]

#(1) Exercice of the following task
results = list(apriori(transactions_bike_shop, min_support=0.15, min_confidence=0.9, min_lift=1.0))
print_apyori_output(results)

Rules involving itemset ['sun_glasses', 'road_bike', 'helmet']
['road_bike', 'helmet'] => ['sun_glasses'] (support=0.1500, confidence=1.00, lift=2.86)

Rules involving itemset ['sun_glasses', 'water_bottle', 'helmet']
['water_bottle', 'helmet'] => ['sun_glasses'] (support=0.2000, confidence=1.00, lift=2.86)
['sun_glasses', 'water_bottle'] => ['helmet'] (support=0.2000, confidence=1.00, lift=1.67)



<font size="+1" color="red">Replace this cell with a markdown cell containing (1) a printout of the rules you have obtained, and (2) for each of those rules, indicate clearly how the support, confidence, and lift is calculated. Do not merely repeat the formula: indicate how each number is computed based on the transactions you provided, as if you were trying to verify that the numbers are correct.</font>

<font size="+1" color="blue">(1) printed above
    (2) For each rule:
    **(2.1)['helmet', 'road_bike'] => ['sun_glasses']
    (2.2)['helmet', 'water_bottle'] => ['sun_glasses']
    (2.3)['water_bottle', 'sun_glasses'] => ['helmet']**
    
    In (2.1) the sup( ['helmet', 'road_bike'] ) is the #transactions [helmet, road_bike] (3)  divided by the total number of transactions (20). So, sup(['helmet', 'sun_glasses'])=3/20 =0.15
    
    The confidence is #transactions containing ['helmet', 'road_bike', 'sun_glasses'] (3) divided by the #transactions of ['helmet', 'road_bike'] (3). So, conf(['helmet', 'road_bike'] => ['sun_glasses'])=3/3=1
    The lift is the confidence of ['helmet', 'road_bike'] => ['sun_glasses']  (1) divided by support of ['sun_glasses'] 
    So, we have: lift(['helmet', 'road_bike'] => ['sun_glasses'])=1/sup(['sun_glasses'])=1/(7/20)=1/0.35=2.857
    
    
    
    
    Now for (2.2) **['helmet', 'water_bottle'] => ['sun_glasses']**
    the sup(['helmet', 'water_bottle']) is the #transactions ['helmet', 'water_bottle'] (4)  divided by the total number of transactions (20). So, sup(['helmet', 'water_bottle'])=4/20 =0.2
    
     The confidence is #transactions containing ['helmet', 'sun_glasses', 'water_bottle'] (4) divided by the #transactions of ['helmet', 'water_bottle'] (4). So, conf(['helmet', 'water_bottle'] => ['sun_glasses'])=4/4=1
    
    So, we have: lift(['helmet', 'water_bottle'] => ['sun_glasses'])=conf(['helmet', 'water_bottle'] => ['sun_glasses'])/sup(['sun_glasses'])=1/(7/20)=1/0.35=2.857
    
    
        Now for (2.3) **['water_bottle', 'sun_glasses'] => ['helmet']**
    the sup(['water_bottle', 'sun_glasses'] ) is the #transactions ['water_bottle', 'sun_glasses'] (4)  divided by the total number of transactions (20). So, sup(['water_bottle', 'sun_glasses'])=4/20 =0.2
    
     The confidence is #transactions containing ['helmet', 'sun_glasses', 'water_bottle'] (4) divided by the #transactions of ['water_bottle', 'sun_glasses'] (4). So, conf(['water_bottle', 'sun_glasses'] => ['helmet'])=4/4=1
    
    So, we have: lift(['water_bottle', 'sun_glasses'] => ['helmet'])=conf(['water_bottle', 'sun_glasses'] => ['helmet'])/sup(['helmet'])=1/(12/20)=1/0.6=1.67
    
    
    After all, I've verified that the results is correct
    
   </font>

# 2. Load and prepare the shopping baskets

In [4]:
# LEAVE AS-IS

# File names
INPUT_PRODUCTS = "instacart-products.csv"
INPUT_TRANSACTIONS = "instacart-transactions.csv.gz"

# Read into a dataframe
products = pd.read_csv(INPUT_PRODUCTS, delimiter=",")

# Set product_id as index, and drop column aisle_id
products = products.set_index('product_id').drop(columns=['aisle_id'])

products.head(100)

product_name  department_id
product_id                                                                  
1                                  Chocolate Sandwich Cookies             19
2                                            All-Seasons Salt             13
3                        Robust Golden Unsweetened Oolong Tea              7
4           Smart Ones Classic Favorites Mini Rigatoni Wit...              1
5                                   Green Chile Anytime Sauce             13
...                                                       ...            ...
96                    Sprinklez Confetti Fun Organic Toppings             13
97                                Organic Chamomile Lemon Tea              7
98                                  2% Yellow American Cheese             16
99                                Local Living Butter Lettuce              4
100                   Peanut Butter & Strawberry Jam Sandwich              1

[100 rows x 2 columns]

## 2.1. Select by department

In [5]:
# LEAVE AS-IS

DEPT_BAKERY = 3
DEPT_VEGGIES = 4
DEPT_ALCOHOL = 5
DEPT_WORLD = 6
DEPT_DRINKS = 7
DEPT_PETS = 8
DEPT_PHARMACY = 11
DEPT_CLEANING = 17
DEPT_BABIES = 18

<font size="+1" color="red">Replace this cell with your code for *select_from_departments*.</font>

In [6]:
#code for select_from_departments.
def select_from_departments(df,product_ids,department_ids):
    
    list_products_id=[]#definition of the list I'll return
    for product in product_ids:
        if df.loc[product].department_id in department_ids:#if the department_id of a product is in department_ids
            list_products_id.append(product)   #then we add this product to the list that we'll have to return
    
    return list_products_id
    

<font size="+1" color="red">Replace this cell with code to test your function with three different test cases. Each test case is a list of items and a list of 1, 2, or 3 departments.</font>

In [7]:
def testing(tc,department):#create the a funciton to avoid repeating code
    test_case=[]
    test_case=tc #copy the array to test_case
    print("Test case:")#print the test case
    print(test_case)

    print("\nInput products:")#print the input products as the model
    for product_id in test_case:
        print(f"{product_id} {products.loc[product_id].product_name} (dept {products.loc[product_id].department_id})")


    print("\nSelected products:")#print the products in which the department of the product coincides with the department
    selected_products=[]
    selected_products=select_from_departments(products,test_case,department)
    for i in selected_products:
        print(f"{i} {products.loc[i].product_name} (dept {products.loc[i].department_id})")


In [8]:
test_case=[[22, 26, 45, 54, 57, 71, 111, 1120],
           [1,2,3,4,5,6,7,8,9,10,11,12],
           [13,14,15,16,17,18,19,20,21,2200,23,24],
          ]
departments=[[17],[11,19],[1,12,19]]
for i in range(3):
    print()
    testing(test_case[i],departments[i])


Test case:
[22, 26, 45, 54, 57, 71, 111, 1120]

Input products:
22 Fresh Breath Oral Rinse Mild Mint (dept 11)
26 Fancy Feast Trout Feast Flaked Wet Cat Food (dept 8)
45 European Cucumber (dept 4)
54 24/7 Performance Cat Litter (dept 8)
57 Flat Toothpicks (dept 17)
71 Ultra 7 Inch Polypropylene Traditional Plates (dept 17)
111 Fabric Softener, Geranium Scent (dept 17)
1120 All American Flame Grilled Veggie Burgers (dept 1)

Selected products:
57 Flat Toothpicks (dept 17)
71 Ultra 7 Inch Polypropylene Traditional Plates (dept 17)
111 Fabric Softener, Geranium Scent (dept 17)

Test case:
[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]

Input products:
1 Chocolate Sandwich Cookies (dept 19)
2 All-Seasons Salt (dept 13)
3 Robust Golden Unsweetened Oolong Tea (dept 7)
4 Smart Ones Classic Favorites Mini Rigatoni With Vodka Cream Sauce (dept 1)
5 Green Chile Anytime Sauce (dept 13)
6 Dry Nose Oil (dept 11)
7 Pure Coconut Water With Orange (dept 7)
8 Cut Russet Potatoes Steam N' Mash (dept 1)
9 Ligh

## 2.2. Read and filter transactions

<font size="+1" color="red">Replace this cell with your code to read transactions, keeping only items in DEPT_PHARMACY. Remember to stop after storing 5000 of the transactions read.</font>

In [9]:
#code to read transactions, keeping only items in DEPT_PHARMACY. Remember to stop after storing 5000 of the transactions read.



def TRANSACTIONS(department): #I created this function that generalizes the code to get the transactions of a department
    #because in the last exercice, they ask to do the same but for other departments. So I thought that I would define it here above and I can already start simplifying code
    transactions=[]
    with gzip.open(INPUT_TRANSACTIONS, "rt") as inputfile:

        # Create a CSV reader
        reader = csv.reader(inputfile, delimiter=",")


        count=0#count the transactions stored
        # Iterate through the CSV fil

        for row in reader:
            if count==5000:
                print("break")
                break

            # Convert to integers
            items = [int(x) for x in row]
            if isinstance(department, int):
                department = [department]
            dept_items=select_from_departments(products,items,department)
            
            #item_check=0
            #for item in items:#we check if there's any item in items that its department is the pharmacy
             #   if products.loc[item,"department_id"]==departament:
              #      item_check=1
            if dept_items:#if there's any item of pharmacy in the row (items)
                transactions.append(dept_items)#then we append the items into the array transactions
                count += 1
                if (count>0  and  count% 1000 == 0):
                    print(f"transactions read: {reader.line_num}, transactions stored: {count}")
    
    return transactions


In [10]:
#code to read transactions, keeping only items in DEPT_PHARMACY. Remember to stop after storing 5000 of the transactions read.
transaction=[]
transaction=TRANSACTIONS(DEPT_PHARMACY)


transactions read: 9620, transactions stored: 1000
transactions read: 19479, transactions stored: 2000
transactions read: 29003, transactions stored: 3000
transactions read: 38755, transactions stored: 4000
transactions read: 48881, transactions stored: 5000
break


## 2.3. Extract association rules and comment on them

<font size="+1" color="red">Replace this cell with your code to extract association rules from the read transactions.</font>

In [11]:
def function_print_rules(minsup,minconf,minlift,transactions):
    results = list(apriori(transactions, min_support=minsup, min_confidence=minconf, min_lift=minlift))
    print_apyori_output(results, products, 'product_name')

#I've created this function with the same purpose as the function above on transactions, because below we are asked to repeat code
#code to extract association rules from the read transactions.
#I define the attributes
minsup=0.001
minconf=0.1
minlift=1
#Rules for DEPT_BAKERY
print("Rules for DEPT_PHARMACY")
function_print_rules(minsup,minconf,minlift,transaction)

Rules for DEPT_PHARMACY
Rules involving itemset [16825, 1406]
['Vegan Nutritional Shake Sweet Vanilla Bean'] => ['Vegan Smooth Chocolate Nutritional Shake'] (support=0.0012, confidence=0.40, lift=80.00)
['Vegan Smooth Chocolate Nutritional Shake'] => ['Vegan Nutritional Shake Sweet Vanilla Bean'] (support=0.0012, confidence=0.24, lift=80.00)

Rules involving itemset [7976, 16019]
['One Plant-Based Chocolate Flavor Nutritional Shake Drink Mix'] => ['All-In-One French Vanilla Nutritional Shake Sachet'] (support=0.0010, confidence=0.31, lift=97.66)
['All-In-One French Vanilla Nutritional Shake Sachet'] => ['One Plant-Based Chocolate Flavor Nutritional Shake Drink Mix'] (support=0.0010, confidence=0.31, lift=97.66)

Rules involving itemset [27544, 39162]
['Lemon Verbena Hand Soap'] => ['Clean Day Basil Hand Soap'] (support=0.0018, confidence=0.12, lift=11.21)
['Clean Day Basil Hand Soap'] => ['Lemon Verbena Hand Soap'] (support=0.0018, confidence=0.16, lift=11.21)



<font size="+1" color="red">Replace this cell with a brief commentary on what you would recommend to the shopping app considering the extracted association rules.</font>

<font size="+1" color="blue">After printing and analysing the extracted association rules, I would recommend to the shopping app the following: 
    
    Usually, nutritional products are bought together. That means that if a customer buys an nutritional product such as 'Vegan Nutritional Shake', it is very likely that will buy also other products similar to it.
    
    Another point to consider, Nutritional Shakes are usually bought together with other shakes, as well as Soaps with other soaps.    
    That means that if a customer buys any of those products, I would recommend to the application that it shows also the products that are related, as it is very likely, that the customer has also in mind to buy some of those.
</font>

## 2.4. Extract association rules and comment on them (other departments)

<font size="+1" color="red">Replace this cell with code to select a different set of departments (at least two, not DEPT_PHARMACY) and extract transactions again. Avoid replicating code when possible.</font>

In [12]:
#code to select a different set of departments (at least two, not DEPT_PHARMACY) and extract transactions again. 
departments_array=[DEPT_BAKERY,DEPT_VEGGIES]#I create and array to keep the departments

#I create a funcion called TRANSACTIONS that generalizes the transactions of a department
print("Print transactions DEPT_BAKERY: ")
transaction1=TRANSACTIONS(departments_array[0])

print("Print transactions DEPT_VEGGIES: ")
transaction2=TRANSACTIONS(departments_array[1])
print()

#I define each parameter to pass it to the next function that I created, to avoid code repetition
#Simply the function function_print_rules, prints the rules, given the input parameters of below and a transaction array


                


Print transactions DEPT_BAKERY: 
transactions read: 3546, transactions stored: 1000
transactions read: 7226, transactions stored: 2000
transactions read: 10888, transactions stored: 3000
transactions read: 14527, transactions stored: 4000
transactions read: 18321, transactions stored: 5000
break
Print transactions DEPT_VEGGIES: 
transactions read: 1298, transactions stored: 1000
transactions read: 2638, transactions stored: 2000
transactions read: 3977, transactions stored: 3000
transactions read: 5334, transactions stored: 4000
transactions read: 6684, transactions stored: 5000
break



In [13]:
#continuation of the code of above

minsup=0.001
minconf=0.1
minlift=1
#Rules for DEPT_BAKERY
print("Rules for DEPT_BAKERY")
function_print_rules(minsup,minconf,minlift,transaction1)
print()
minsup=0.02
#Rules for DEPT_VEGGIES
print("Rules for DEPT_VEGGIES")
function_print_rules(minsup,minconf,minlift,transaction2)

Rules for DEPT_BAKERY
Rules involving itemset [10673, 5077]
['Original Nooks & Crannies English Muffins'] => ['100% Whole Wheat Bread'] (support=0.0018, confidence=0.10, lift=1.40)

Rules involving itemset [17706, 5077]
['Organic Whole Grain Wheat English Muffins'] => ['100% Whole Wheat Bread'] (support=0.0018, confidence=0.10, lift=1.44)

Rules involving itemset [42460, 5077]
['Plain Mini Bagels'] => ['100% Whole Wheat Bread'] (support=0.0010, confidence=0.14, lift=1.89)

Rules involving itemset [46842, 5077]
['Plain Pre-Sliced Bagels'] => ['100% Whole Wheat Bread'] (support=0.0010, confidence=0.10, lift=1.43)

Rules involving itemset [5077, 48183]
['Flour Tortillas'] => ['100% Whole Wheat Bread'] (support=0.0010, confidence=0.19, lift=2.59)

Rules involving itemset [10673, 17341]
['Presliced Everything Bagels'] => ['Original Nooks & Crannies English Muffins'] (support=0.0014, confidence=0.15, lift=8.27)

Rules involving itemset [20738, 16262]
['Everything Bagels'] => ['Plain Bagels']

<font size="+1" color="red">Replace this cell with your commentary on the obtained rules.</font>

<font size="+1" color="blue">After printing and analysing the extracted association rules, I would recommend to the shopping app the following: 
    **Concerning the DEPT_BAKERY:**
    I can observe that '100% Whole Wheat Bread' is a very common product bought in the Bakery, as for a lot of people its one of the bases of its diet.
    As well as the '100% Whole Wheat Bread', different types of bagels are frequently bought together with other bagels
    Then, I would recommend to the application that if any costumer buys a product in the bakery, it should recommend the '100% Whole Wheat Bread' as it is commonly bought together. Also, if a customer buys any type of bagel, then it would be appropriate to recommend it to him/her other types of bagels, as might be also interested in.
    **Concerning the DEPT_VEGGIES:**
    Usually, "Organic" products are bought together. That means that if a customer buys an organic product, it is very likely that will buy also other "Organic" items.
    Another point to consider, 'Bag of Organic Bananas' are a common complement to those products bought and that are organic and "Organic" fruits.
    That means that if a customer buys any of those products, such as Fruits, or Organic items I would recommend to the application that it shows also the 'Bag of Organic Bananas' and other organic product, as it is very likely, that the customer has also in mind to buy those.
</font>

<font size="+2" color="#003300">I hereby declare that, except for the code provided by the course instructors, all of my code, report, and figures were produced by myself.</font>